src/coreclr/jit/fgprofile.cpp

```diff
@@ -439,443 +447,1176 @@ void BlockCountInstrumentor::InstrumentMethodEntry(Schema& schema, BYTE* profile
+                // preferentially visit critical edges first; if these
+                // edges end up in the DFST then instrumentation will
+                // require edge splitting.
+                //
+                // We also want to preferentially visit edges to rare
+                // successors last, if this block is non-rare.
+                //
+                // It's not immediately clear if we should pass comp or this
+                // to NumSucc here (for inlinees).
+                //
+                // It matters for FINALLYRET and for SWITCHES. Currently
+                // we handle the first one specially, and it seems possible
+                // things will just work for switches either way, but it
+                // might work a bit better using the root compiler.
+                //
+                const unsigned numSucc = block->NumSucc(comp);
 
-public:
-    ClassProbeInserter(Schema& schema, BYTE* profileMemory, int* pCurrentSchemaIndex, unsigned& instrCount)
-        : m_schema(schema)
-        , m_profileMemory(profileMemory)
-        , m_currentSchemaIndex(pCurrentSchemaIndex)
-        , m_instrCount(instrCount)
-    {
-    }
+                if (numSucc == 1)
+                {
+                    // Not a fork. Just visit the sole successor.
+                    //
+                    BasicBlock* const target = block->GetSucc(0, comp);
+                    if (BlockSetOps::IsMember(comp, marked, target->bbNum))
+                    {
+                        // We can't instrument in the call always pair tail block
+                        // so treat this as a critical edge.
+                        //
+                        visitor->VisitNonTreeEdge(block, target, block->isBBCallAlwaysPairTail()
+                                                                     ? SpanningTreeVisitor::Kind::Edge
+                                                                     : SpanningTreeVisitor::Kind::Source);
+                    }
+                    else
+                    {
+                        visitor->VisitTreeEdge(block, target);
+                        stack.Push(target);
+                        BlockSetOps::AddElemD(comp, marked, target->bbNum);
+                    }
+                }
+                else
+                {
+                    // A block with multiple successors.
+                    //
+                    // Because we're using a stack up above, we work in reverse
+                    // order of "cost" here --  so we first consider rare,
+                    // then normal, then critical.
+                    //
+                    // That is, all things being equal we'd prefer to
+                    // have critical edges be tree edges, and
+                    // edges from non-rare to rare be non-tree edges.
+                    //
+                    scratch.Reset();
+                    BlockSet processed = BlockSetOps::MakeEmpty(comp);
```

Comment:
throughput nit: maybe this should be allocated outside the loop and reset here?

Reply:
Not sure it saves much..? But now initialize above and clear here.